### PR TITLE
finished Projects

### PR DESCRIPTION
**Pull request by 0xf24**

### CODE REVIEW

File: ReportCardParser/src/reportcardparser/ReportCardParserType3.java [hunk 1]
```diff
@@ -10,28 +10,63 @@
...
+    public ReportCard parse(String[] lines) {
+        
+        //get the name by splitting the first line by ':' and trimming the second element
+        String name = lines[0].split(":")[1].trim();
+        
+        //get the date of birth by splitting the second line by ':' and trimming the second element
+        String dateOfBirth = lines[1].split(":")[1].trim();
+        
+        //get the course names from the 4th line, splitting by tabs
+        String[] courseNames = lines[3].split("([\\t]+)");
```
Comment (cgallinaro) [inline ReportCardParser/src/reportcardparser/ReportCardParserType3.java:47]: Good use of string splitting.

File: ReportCardParser/src/reportcardparser/ReportCardStats.java [hunk 2]
```diff
@@ -42,7 +66,33 @@ public static String GetBestSubject(ReportCard reportCard) {
...
+        return marksSum / reportCards.length;
+    }
+    
+    /**
+     * Get a CourseMark by name from a report card
+     * @param card the card which has the marks
+     * @param subject the name of the subject to get
+     * @return the CourseMark of the subject
+     */
+    public static CourseMark getSubjectMark(ReportCard card, String subject) {
```
Comment (cgallinaro) [inline ReportCardParser/src/reportcardparser/ReportCardStats.java:90]: Your getSubjectMark helper method can return null, but you don’t check for this (or write a test for it).

File: ReportCardParser/src/reportcardparser/ReportCardStats.java [hunk 3]
```diff
@@ -19,8 +19,20 @@
      * @return The name of the student with the highest average
      */
     public static String GetBestStudent(ReportCard[] reportCards) {
-        // REPLACE ME WITH A REAL IMPLEMENTATION!
-        return null;
+        
+        //store the current best reportCard
+        ReportCard bestAverage = reportCards[0];
```
Comment (cgallinaro) [inline ReportCardParser/src/reportcardparser/ReportCardStats.java:24]: What if reportCards is null?  What if the array entries are null?

File: ReportCardParser/src/reportcardparser/ReportCardStats.java [hunk 4]
```diff
@@ -30,8 +42,20 @@ public static String GetBestStudent(ReportCard[] reportCards) {
      * @return The name of the subject with the highest mark
      */
     public static String GetBestSubject(ReportCard reportCard) {
-        // REPLACE ME WITH A REAL IMPLEMENTATION!
-        return null;
+        
+        //store the current best mark (subject)
+        CourseMark bestMark = reportCard.getMarks()[0];
```
Comment (cgallinaro) [inline ReportCardParser/src/reportcardparser/ReportCardStats.java:47]: What if reportCard is null?

File: ReportCardParser/src/reportcardparser/ReportCardStats.java [hunk 5]
```diff
@@ -42,7 +66,33 @@ public static String GetBestSubject(ReportCard reportCard) {
...
      */
     public static double SubjectAverage(ReportCard[] reportCards, String subject) {
-        // REPLACE ME WITH A REAL IMPLEMENTATION!
-        return 0.0;
+        
+        //sum of all the marks of the subject
+        double marksSum = 0;
+        
+        //ittorate through all of the reportCards
+        for (ReportCard card : reportCards) {
```
Comment (cgallinaro) [inline ReportCardParser/src/reportcardparser/ReportCardStats.java:74]: What if reportCards is null (or the array entries are null)?  What if the subject is not found for a particular student?

File: ReportCardParser/test/reportcardparser/ReportCardStatsTest.java [hunk 6]
```diff
@@ -5,7 +5,8 @@
  */
 package reportcardparser;
 
-import org.junit.jupiter.api.Test;
+import static org.junit.Assert.*;
```
Comment (cgallinaro) [inline ReportCardParser/test/reportcardparser/ReportCardStatsTest.java:8]: The tests you have written are good, but you are missing tests for edge/error cases.

File: ReportCardParser/src/reportcardparser/AbstractReportCardParser.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        }
+        
+        //split contents into an array of lines, split using the system's newline character
+        String[] lines = contents.split(System.lineSeparator());
+        
+        //return the implementers parse method with the separated lines
+        return this.parse(lines);
+    }
+    
+    public abstract ReportCard parse(String[] lines);
```
Comment (cgallinaro) [inline ReportCardParser/src/reportcardparser/AbstractReportCardParser.java:56]: You abstract method should be protected (not public).  We don’t want users of the concrete parsing classes to be able to call this abstract method.  This abstract method is also missing documentation.

File: TicTacToe/src/tictactoe/TicTacToe.java [hunk 8]
```diff
@@ -4,7 +4,10 @@
     
     private char[][] board;
 
-    // Constructors
+    /**
```
Comment (cgallinaro) [inline TicTacToe/src/tictactoe/TicTacToe.java:7]: Missing documentation at top of class and documentation of board member variable.

File: TicTacToe/test/tictactoe/TicTacToeTest.java [hunk 9]
```diff
@@ -7,30 +7,168 @@
...
+        
+        assertEquals(testBoard.isCat(), false);
+        
+        testBoard = new TicTacToe(new char[][] {
+            {X, O, X},
+            {O, O, X},
+            {X, X, E}
+        });
+        
+        assertEquals(testBoard.isCat(), false);
```
Comment (cgallinaro) [inline TicTacToe/test/tictactoe/TicTacToeTest.java:112]: No tests for isCat == true.  No checks for when the board is full but there is a winner (which means isCat would be false).

File: TicTacToe/test/tictactoe/TicTacToeTest.java [hunk 10]
```diff
@@ -7,30 +7,168 @@
...
+            {X, O, E},
+            {O, E, X},
+            {X, X, E}
+        });
+                        
+        assertEquals(testBoard.isValid(2, 2), true);
+        
+        assertEquals(testBoard.isValid(0, 1), true);
+        
+        assertEquals(testBoard.isValid(3, 2), false);
```
Comment (cgallinaro) [inline TicTacToe/test/tictactoe/TicTacToeTest.java:128]: Should cover all valid moves.

File: TicTacToe/test/tictactoe/TicTacToeTest.java [hunk 11]
```diff
@@ -7,30 +7,168 @@
...
-    public void testDisplayBoard() {
+        TicTacToe testBoard = new TicTacToe(new char[][] {
+            {X, O, E},
+            {O, E, X},
+            {X, X, E}
+        });
+        
+        assertEquals(testBoard.playerAt(0, 0), X);
+        assertEquals(testBoard.playerAt(2, 1), X);
+        assertEquals(testBoard.playerAt(0, 1), O);
```
Comment (cgallinaro) [inline TicTacToe/test/tictactoe/TicTacToeTest.java:142]: No test for empty space.  No test for invalid move.

File: TicTacToe/src/tictactoe/TicTacToeAI.java [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+ * To change this template file, choose Tools | Templates
+ * and open the template in the editor.
+ */
+package tictactoe;
+
+/**
+ *
+ * @author manxboy
+ */
+public class TicTacToeAI {
```
Comment (cgallinaro) [inline TicTacToe/src/tictactoe/TicTacToeAI.java:12]: The player is supposed to play as X (and go first).  In your game, the AI plays as X.

**Comment by cgallinaro** [comment]
Problem 1:
Implementation of parsers: 5/5
Implementation of ReportCardStats methods and tests: 8/10
Implementation of AbstractReportCardParser: 4.5/5

Problem 2:
Documentation: 1.5/2
TicTacToe implementation: 5/5
TicTacToeTest implementation: 4/5
TicTacToeAI implementation: 7/8

Total: 35/40
